packages/router/utils.ts

```diff
@@ -415,48 +419,98 @@ function flattenRoutes<
       return;
     }
 
-    // Handle optional params - /path/:optional?
-    let segments = path.split("/");
-    let optionalParams: string[] = [];
-    segments.forEach((segment) => {
-      let match = segment.match(/^:?([^?]+)\?$/);
-      if (match) {
-        optionalParams.push(match[1]);
-      }
+    branches.push({
+      path,
+      score: computeScore(path, route.index),
+      routesMeta,
     });
-
-    if (optionalParams.length > 0) {
-      for (let i = 0; i <= optionalParams.length; i++) {
-        let newPath = path;
-        let newMeta = routesMeta.map((m) => ({ ...m }));
-
-        for (let j = optionalParams.length - 1; j >= 0; j--) {
-          let re = new RegExp(`(\\/:?${optionalParams[j]})\\?`);
-          let replacement = j < i ? "$1" : "";
-          newPath = newPath.replace(re, replacement);
-          newMeta[newMeta.length - 1].relativePath = newMeta[
-            newMeta.length - 1
-          ].relativePath.replace(re, replacement);
-        }
-
-        branches.push({
-          path: newPath,
-          score: computeScore(newPath, route.index),
-          routesMeta: newMeta,
-        });
-      }
+  };
+  routes.forEach((route, index) => {
+    // coarse-grain check for optional params
+    if (route.path === "" || !route.path?.includes("?")) {
+      flattenRoute(route, index);
     } else {
-      branches.push({
-        path,
-        score: computeScore(path, route.index),
-        routesMeta,
-      });
+      for (let exploded of explodeOptionalSegments(route.path)) {
+        flattenRoute(route, index, exploded);
+      }
```

Comment:
I think this is close - but I'm still seeing some `?`'s in flattened paths (and also in the then-generated `routesMeta.relativePath`.  I think always as part of the last optional segment?  Here's the flattened/ranked paths for the unit test above:

<img width="557" alt="Screenshot 2022-12-06 at 6 22 02 PM" src="https://user-images.githubusercontent.com/1609022/206046796-1b6df323-4dce-4a08-ad6f-823976a9d9db.png">

Otherwise the `routesMeta` relative paths are looking right to me though!

Reply:
I'll try to hit this with a few more complex unit tests scenarios tomorrow as well 👍

Reply:
Yes, forgot to update code to handle empty segments for nested routes. Should work now!